File: packages/cli/src/commands/audit.ts
import chalk from "chalk"
import { loadPalamedesConfig, type LoadedPalamedesConfig } from "@palamedes/config"
import {
  auditCatalogs,
  type CatalogAuditDiagnostic,
  type CatalogAuditResult,
} from "@palamedes/core-node"

type AuditOptions = {
  config?: string
  locale?: string[]
  json?: boolean
  failOn?: "error" | "warning"
}

export async function audit(options: AuditOptions): Promise<CatalogAuditResult> {
  const config = await loadPalamedesConfig({ configPath: options.config })
  const result = auditCatalogs(toAuditConfig(config), {
    locales: options.locale,
  })

  if (options.json) {
    console.log(JSON.stringify(result, null, 2))
  } else {
    printAuditResult(result)
  }

  const failOn = normalizeFailOn(options.failOn)
  if (shouldFail(result, failOn)) {
    throw new Error(createFailureMessage(result, failOn))
  }

  return result
}

function toAuditConfig(config: LoadedPalamedesConfig) {
  return {
    rootDir: config.rootDir,
    locales: config.locales,
    sourceLocale: config.sourceLocale,
    fallbackLocales: config.fallbackLocales,
    pseudoLocale: config.pseudoLocale,
    catalogs: config.catalogs,
  }
}

function printAuditResult(result: CatalogAuditResult): void {
  const { summary } = result
  const status = summary.errors > 0 ? chalk.red("failed") : chalk.green("passed")
  console.log(
    `Catalog audit ${status}: ${summary.errors} error(s), ${summary.warnings} warning(s), ${summary.infos} info`
  )

  if (result.diagnostics.length === 0) {
    return
  }

  for (const [catalogPath, diagnostics] of groupByCatalog(result.diagnostics)) {
    console.log(chalk.gray(`\n${catalogPath}`))
    for (const diagnostic of diagnostics) {
      const severity = formatSeverity(diagnostic.severity)
      const locale = diagnostic.locale ? ` ${chalk.gray(`[${diagnostic.locale}]`)}` : ""
      const source = formatSource(diagnostic)
      console.log(`  ${severity} ${diagnostic.code}${locale}: ${diagnostic.message}${source}`)
    }
  }
}

function groupByCatalog(diagnostics: CatalogAuditDiagnostic[]) {
  const grouped = new Map<string, CatalogAuditDiagnostic[]>()
  for (const diagnostic of diagnostics) {
    const entries = grouped.get(diagnostic.catalogPath) ?? []
    entries.push(diagnostic)
    grouped.set(diagnostic.catalogPath, entries)
  }
  return grouped
}

function formatSeverity(severity: CatalogAuditDiagnostic["severity"]): string {
  switch (severity) {
    case "error": {
      return chalk.red("[error]")
    }
    case "warning": {
      return chalk.yellow("[warning]")
    }
    case "info": {
      return chalk.blue("[info]")
    }
  }
}

function formatSource(diagnostic: CatalogAuditDiagnostic): string {
  if (!diagnostic.sourceKey) {
    return ""
  }

  const context = diagnostic.sourceKey.context ? ` [context: ${diagnostic.sourceKey.context}]` : ""
  return chalk.gray(`\n    Source: ${diagnostic.sourceKey.message}${context}`)
}

function shouldFail(result: CatalogAuditResult, failOn: "error" | "warning"): boolean {
  if (failOn === "warning") {
    return result.summary.errors > 0 || result.summary.warnings > 0
  }

  return result.summary.errors > 0
}

function normalizeFailOn(value: AuditOptions["failOn"]): "error" | "warning" {
  if (value === undefined || value === "error" || value === "warning") {
    return value ?? "error"
  }

  throw new Error("Invalid --fail-on value. Expected `error` or `warning`.")
}

function createFailureMessage(result: CatalogAuditResult, failOn: "error" | "warning"): string {
  if (failOn === "warning") {
    return `Catalog audit failed with ${result.summary.errors} error(s) and ${result.summary.warnings} warning(s).`
  }

  return `Catalog audit failed with ${result.summary.errors} error(s).`
}

File: crates/palamedes-cli/src/commands/audit.rs
//! `pmds audit` — translation and ICU authoring checks across catalogs.

use std::collections::BTreeMap;
use std::path::PathBuf;

use clap::{Args, ValueEnum};
use palamedes::{
    audit_catalogs, CatalogAuditDiagnostic, CatalogAuditRequest, CatalogAuditResult,
    CatalogAuditSummary,
};

use crate::command::{render_json, Command, Context};
use crate::commands::normalize_locale_list;
use crate::error::CliError;

#[derive(Debug, Args)]
pub struct AuditOptions {
    /// Path to a Palamedes config file.
    #[arg(short, long)]
    config: Option<PathBuf>,
    /// Only audit selected target locale(s).
    #[arg(long, num_args = 1..)]
    locale: Vec<String>,
    /// Print the machine-readable audit result as JSON.
    #[arg(long)]
    json: bool,
    /// Fail on error, warning, or info diagnostics.
    #[arg(long, default_value = "error")]
    fail_on: FailOn,
}

#[derive(Clone, Copy, Debug, ValueEnum)]
enum FailOn {
    Error,
    Warning,
    Info,
}

impl Command for AuditOptions {
    type Output = CatalogAuditResult;

    fn run(&self, context: &Context) -> Result<Self::Output, CliError> {
        let config = context.load_config(self.config.as_deref())?;
        Ok(audit_catalogs(CatalogAuditRequest {
            config: config.artifact_config(),
            locales: normalize_locale_list(&self.locale),
            checks: Default::default(),
            metadata: Vec::new(),
        })?)
    }

    fn render(&self, output: &Self::Output) -> Result<(), CliError> {
        if self.json {
            return render_json(output);
        }
        print_audit_result(output, self.fail_on);
        Ok(())
    }

    fn verdict(&self, output: &Self::Output) -> Result<(), CliError> {
        if !self.fail_on.should_fail(&output.summary) {
            return Ok(());
        }

        match self.fail_on {
            FailOn::Info => Err(CliError::AuditFailedOnInfo {
                errors: output.summary.errors,
                warnings: output.summary.warnings,
                infos: output.summary.infos,
            }),
            FailOn::Warning => Err(CliError::AuditFailedOnWarning {
                errors: output.summary.errors,
                warnings: output.summary.warnings,
            }),
            FailOn::Error => Err(CliError::AuditFailedOnError {
                errors: output.summary.errors,
            }),
        }
    }
}

impl FailOn {
    fn should_fail(self, summary: &CatalogAuditSummary) -> bool {
        match self {
            Self::Info => summary.errors > 0 || summary.warnings > 0 || summary.infos > 0,
            Self::Warning => summary.errors > 0 || summary.warnings > 0,
            Self::Error => summary.errors > 0,
        }
    }
}

fn print_audit_result(result: &CatalogAuditResult, fail_on: FailOn) {
    let status = if fail_on.should_fail(&result.summary) {
        "failed"
    } else {
        "passed"
    };
    println!(
        "Catalog audit {status}: {} error(s), {} warning(s), {} info",
        result.summary.errors, result.summary.warnings, result.summary.infos
    );

    let mut grouped = BTreeMap::<&str, Vec<&CatalogAuditDiagnostic>>::new();
    for diagnostic in &result.diagnostics {
        grouped
            .entry(diagnostic.catalog_path.as_str())
            .or_default()
            .push(diagnostic);
    }

    for (catalog_path, diagnostics) in grouped {
        println!("\n{catalog_path}");
        for diagnostic in diagnostics {
            let locale = diagnostic
                .locale
                .as_ref()
                .map(|locale| format!(" [{locale}]"))
                .unwrap_or_default();
            println!(
                "  [{:?}] {}{}: {}",
                diagnostic.severity, diagnostic.code, locale, diagnostic.message
            );
            if let Some(source_key) = &diagnostic.source_key {
                let context = source_key
                    .context
                    .as_ref()
                    .map(|context| format!(" [context: {context}]"))
                    .unwrap_or_default();
                println!("    Source: {}{}", source_key.message, context);
            }
        }
    }
}

#[cfg(test)]
mod tests {
    use super::FailOn;
    use palamedes::CatalogAuditSummary;

    #[test]
    fn fail_on_thresholds_share_exact_severity_boundaries() {
        let errors = CatalogAuditSummary {
            errors: 1,
            ..CatalogAuditSummary::default()
        };
        let warnings = CatalogAuditSummary {
            warnings: 1,
            ..CatalogAuditSummary::default()
        };
        let infos = CatalogAuditSummary {
            infos: 1,
            ..CatalogAuditSummary::default()
        };

        assert!(FailOn::Error.should_fail(&errors));
        assert!(!FailOn::Error.should_fail(&warnings));
        assert!(FailOn::Warning.should_fail(&warnings));
        assert!(!FailOn::Warning.should_fail(&infos));
        assert!(FailOn::Info.should_fail(&infos));
        assert!(!FailOn::Info.should_fail(&CatalogAuditSummary::default()));
    }
}
